### ISSUE TITLE
Add a small batch quality-gate API with structured results

### ISSUE THREAD

**Issue by AndreaBozzo** [AndreaBozzo/dataprof#376]
Parent: #371

## User problem

The README shows users how to inspect a score, but every CI or ETL user must hand-roll comparisons, missing-metric handling, messages, and exit behavior. #188 concerns continuous streaming alerts; batch profiles still need a small library-level decision API.

## Proposed shape

Provide an explicit policy/result model, for example:

```python
result = report.check(
    min_quality=90,
    max_null_percentage={"customer_id": 0, "*": 20},
    require_metrics=["quality"],
)
if not result.passed:
    for violation in result.violations:
        print(violation.code, violation.message)
```

The exact names should be settled before implementation. The important contract is that evaluation returns structured deterministic data and does not exit the process, print, send alerts, or mutate the report.

## Acceptance criteria

- [ ] Supports an overall quality floor and at least one column-level threshold.
- [ ] Missing/not-analyzed metrics produce an explicit policy outcome; they never pass as zero or fail accidentally due to a default.
- [ ] Violations have stable codes, observed value, expected constraint, and optional column.
- [ ] Results serialize cleanly for CI systems.
- [ ] Rust and Python semantics match for the shared subset.
- [ ] README/example coverage shows an ETL gate without a CLI.
- [ ] Tests cover pass, fail, missing metric, wildcard/column override, and percentage scale (0–100).

## Depends on

- **#715**, hard. This ticket requires that a missing or not-analyzed metric
  produce an explicit policy outcome and never pass as zero or fail by default.
  That is not implementable today: `ReportAssembler::build()` cannot fail, so a
  quality computation that was never requested and one that failed internally
  both arrive as `quality = None`. A gate built on top would give the same
  verdict for "you did not ask for this" and "this broke".

## Relationship to other work

- May consume the structured finding/evidence vocabulary from the findings child of #371, but must not require findings to evaluate numeric policies.
- Does not replace #188, which owns continuous stream triggers and sinks.
- Must preserve `None` = not analyzed versus empty = analyzed with no violations.

## Likely files

- `crates/dataprof-core/src/`
- `crates/dataprof-runtime/src/`
- `python/dataprof/__init__.py` and stubs
- `examples/etl_quality_gate.rs` and Python equivalent
- focused Rust/Python tests

## Non-goals

- process exit codes or CLI flags
- webhooks, logging sinks, or schedulers
- transforming data to make a gate pass

## Evidence scope and incomplete evaluation

A decision needs both its result and the evidence scope supporting it. Current execution metadata already exposes source exhaustion, truncation, and sampling; this API should use that information when a policy asks about the whole source.

- [ ] State whether the decision concerns the scanned rows/selected columns or the full source, and preserve the reason evidence is insufficient.
- [ ] A clean prefix or sample cannot pass a full-source uniqueness requirement. A witnessed duplicate may establish failure for that requirement despite an incomplete scan.
- [ ] Do not generalize sample ratios into full-source pass/fail outcomes without valid bounds. Scope-specific policies can still evaluate the observed sample explicitly.
- [ ] Separate an unrequested or unsupported metric from a completed check with no violations; a requested whole-row check must not appear evaluated after column projection.
- [ ] Cover complete input, capped input, unavailable metrics, projected columns, and a conclusive violation with incomplete coverage in matching Rust/Python fixtures.
- [ ] Preserve verdict, scope, and evidence status/reason when serializing. Keep numeric gate evaluation independent of findings; coordinate vocabulary with #375.

The design motivation is the separation of a verdict from evidence completeness in [mission-level runtime assurance, section 4](https://arxiv.org/html/2607.23532v2). Applying it to profiler policies is a design inference, not a guarantee established by that paper. This does not require a distributed engine or a new global report schema.

**Comment by AndreaBozzo** [comment]
One design note from Databricks' pipeline expectations, which is the closest
shipped thing to this API and gets one part of it wrong in a way worth copying
the opposite of.

Their model has three actions: `warn` (default, invalid records still written),
`drop` (invalid records dropped, count logged with the dataset metrics), and
`fail` (the update does not succeed). The documented consequence of the third is
that **no metrics are recorded at all**, because the update aborts on the first
invalid record. The operator loses the evidence exactly in the case where they
need it most, and has to re-run with a weaker action to find out how bad it was.

For `check()` that translates into a concrete requirement: a failing result must
carry the same evidence a passing one does. `result.passed == False` should
still expose every violation, the observed values, and the constraints that were
met, not a short-circuit at the first breach. The acceptance list already asks
for violations with observed value and expected constraint; this is the case
that would otherwise get optimised away as "we know it failed, stop measuring".

The rest of their taxonomy does not transfer. `drop` and the quarantine pattern
both act on records, and dataprof reports on data rather than moving or filtering
it, so severity (warn versus fail) is the only axis of theirs that belongs here.
That matches #188's `severity` field, and keeping the two vocabularies identical
where they overlap is worth more than adopting the rest.

**Comment by AndreaBozzo** [comment]
## `report.check()` is also what a scikit-learn gate would be built on

Came up while thinking about ML-workflow interoperability. The idea was a
scikit-learn transformer:

```python
Pipeline([
    ("quality", QualityGate(min_quality=90)),
    ("scale", StandardScaler()),
    ("model", LogisticRegression()),
])
```

with `transform(X)` passing `X` through untouched and the profile, violations
and findings exposed as fitted attributes. That is roughly fifty lines over
this issue plus #375, it imports scikit-learn nowhere in the base wheel, and it
does not push dataprof back towards `ml_readiness_score()` or codegen: dataprof
measures, scikit-learn keeps preprocessing and modeling.

I am not proposing it as a child of this issue. It is downstream and would only
be worth packaging once the adapter surface proves durable, probably starting
life in `examples/` rather than as an extra. But it surfaces one design question
that belongs here, and one that does not.

## Belongs here: what a gate does with the schema it did not see

A transformer gates at `fit` on the training split and then sees different data
at every `transform`. Only the first is a batch quality check; the rest are
comparisons against a baseline. That distinction is not sklearn-specific, it is
the difference between "does this batch satisfy a policy" and "does this batch
resemble the one the policy was written against", and it decides whether
`check()` takes only thresholds or can also take a previously computed profile.

Worth settling explicitly, even if the answer is that `check()` only ever
evaluates thresholds and the comparison form belongs to #310's `QualityDelta`.
The failure mode of leaving it implicit is a `check()` that grows a baseline
argument later and has to keep two meanings of "passed".

## Does not belong here

Which frameworks get adapters, and where they live. That is a packaging and
CI-matrix decision, and given #649 the release surface is already the part that
bites. Whatever this issue produces should be usable from a fifty-line adapter
written by anyone, which is a property of the API rather than a list of
supported frameworks.

**Comment by AndreaBozzo** [comment]
## The overall score cannot carry `min_quality`

Measured on current `master`. 500 rows where the key column holds one value 500
times, and every other dimension is clean:

```
order_id,account,amount_eur,booked_at
ORD-1,ACC-0,100.00,2026-08-01
ORD-1,ACC-1,101.00,2026-08-02
...   (498 more, order_id constant)
```

```
overall score: 92.52

uniqueness   key_uniqueness 0.2, duplicate_rows 0
completeness complete_records_ratio 100.0, missing_values_ratio 0.0
consistency  data_type_consistency 100.0, format_violations 0
accuracy     range_violations 0, outlier_ratio 0.0
timeliness   future_dates 0, stale_data_ratio 0.0
validity     valid_values_ratio 100.0
precision    decimal_places_consistency 100.0
```

The primary key is entirely duplicated and the file scores 92.52. It passes
`report.check(min_quality=90)` as written in the issue body.

Two things make it invisible rather than merely discounted:

- Uniqueness carries weight 0.15 (`crates/dataprof-core/src/config.rs:29`), so a
  dimension score of 0 costs at most 15 points. Nothing weighted at 0.15 can
  fail a floor of 90 on its own.
- `duplicate_rows` stays 0, because the full rows differ and only the key
  repeats. The one signal that might have fired independently of the score does
  not fire.

This is not an argument for reweighting. Any weighted mean has the same
property for any dimension: a floor on the aggregate cannot express "this
specific invariant must hold", because averaging is what the aggregate is for.
The score answers "how is this dataset overall", and a gate asks "may this
specific thing through". Those are different questions and #571 already
established that this aggregate is deliberately an average over assessed
dimensions.

## What it implies for the API here

The acceptance criteria say "an overall quality floor and at least one
column-level threshold". The floor is the part that does not work, and the
example in the body leads with it.

Worth deciding explicitly:

- Whether `min_quality` should exist at all in the first iteration, or whether
  shipping it invites exactly the gate above.
- If it ships, whether a violation carries which dimensions were assessed and
  what each scored, so a passing score is auditable rather than a single number.
- Whether the constraint vocabulary covers key uniqueness directly
  (`max_duplicate_key_ratio` or similar), since `key_uniqueness` is already
  computed and is the invariant most gates actually care about. Without it the
  case above has no expressible constraint at all, at any threshold.

The distinction between constraints that can reject and signals that inform is
the same split raised earlier on this issue. This is the concrete case that
shows the aggregate belongs on the signal side.
